app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,13 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def google_oauth2
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      flash[:notice] = I18n.t "devise.omniauth_callbacks.success", kind: "Google"
+      sign_in_and_redirect @user, event: :authentication
+    else
+      session["devise.google_data"] = request.env["omniauth.auth"]
```

Comment:
in .from_omniauth method you find or create user, so there is no need to redirect to new registration page?